File: app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt
package com.waz.zclient.auth

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.waz.zclient.R
import com.waz.zclient.core.config.Config
import com.waz.zclient.core.extension.invisible
import com.waz.zclient.core.extension.startActivityWithAction
import com.waz.zclient.core.extension.visible
import kotlinx.android.synthetic.main.fragment_welcome.*

class WelcomeFragment : Fragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_welcome, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initCreateAccountButtonListener()
        initLoginButtonListener()
        initEnterpriseLoginButtonListener()
        configureEnterpriseLoginVisibility()
    }

    private fun initCreateAccountButtonListener() {
        welcomeCreateAccountButton.setOnClickListener { startCreateAccountFlow() }
    }

    private fun initLoginButtonListener() {
        welcomeLoginButton.setOnClickListener { startLoginFlow() }
    }

    private fun initEnterpriseLoginButtonListener() {
        welcomeEnterpriseLoginButton.setOnClickListener { startEnterpriseLoginFlow() }
    }

    private fun startCreateAccountFlow() {
        startActivityWithAction(ACTION_CREATE_ACCOUNT)
    }

    private fun startLoginFlow() {
        startActivityWithAction(ACTION_LOGIN)
    }

    private fun startEnterpriseLoginFlow() {
        startActivityWithAction(ACTION_SSO_LOGIN)
    }

    private fun configureEnterpriseLoginVisibility() {
        if (Config.allowSso()) welcomeEnterpriseLoginButton.visible()
        else welcomeEnterpriseLoginButton.invisible()
    }
    companion object {
        fun newInstance() = WelcomeFragment()

        private val ACTION_LOGIN = Config.applicationId() + ".LOGIN_ACTION"
        private val ACTION_CREATE_ACCOUNT = Config.applicationId() + ".CREATE_ACCOUNT_ACTION"
        private val ACTION_SSO_LOGIN = Config.applicationId() + ".SSO_LOGIN_ACTION"
    }
}

File: app/src/main/scala/com/waz/zclient/appentry/WelcomeFragment.scala
/**
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.appentry

import android.os.Bundle
import android.view.{LayoutInflater, View, ViewGroup, WindowManager}
import android.widget.Button
import com.waz.utils.returning
import com.waz.zclient._
import com.waz.zclient.appentry.fragments.SignInFragment
import com.waz.zclient.appentry.fragments.SignInFragment.{Email, Login, SignInMethod}
import com.waz.zclient.feature.auth.registration.CreateAccountActivity

class WelcomeFragment extends SSOFragment {

  private lazy val welcomeLoginButton = returning(view[Button](R.id.welcomeLoginButton)) { view =>
    view.onClick { _ =>
      startLoginFlow()
    }
  }

  private lazy val welcomeCreateAccountButton = returning(view[Button](R.id.welcomeCreateAccountButton)) { view =>
    view.onClick { _ =>
      startCreateAccountFlow()
    }
  }

  private lazy val welcomeEnterpriseLoginButton = returning(view[Button](R.id.welcomeEnterpriseLoginButton)) { view =>
    view.foreach(_.setVisibility(if (BuildConfig.ALLOW_SSO) View.VISIBLE else View.INVISIBLE))
    view.onClick { _ =>
      startEnterpriseLoginFlow()
    }
  }

  override def onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle): View =
    inflater.inflate(R.layout.fragment_welcome, container, false)

  override def onViewCreated(view: View, savedInstanceState: Bundle) = {
    super.onViewCreated(view, savedInstanceState)
    initViews()
  }

  override def onResume(): Unit = {
    super.onResume()
    activity.getWindow.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN)
  }

  override def onPause(): Unit = {
    super.onPause()
    activity.getWindow.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
  }

  private def initViews() = {
    welcomeLoginButton
    welcomeCreateAccountButton
    welcomeEnterpriseLoginButton
  }

  private def startCreateAccountFlow(): Unit =
    if (BuildConfig.KOTLIN_REGISTRATION) {
      startActivity(CreateAccountActivity.newIntent(getActivity))
    } else {
      activity.showFragment(CreateAccountFragment(), CreateAccountFragment.Tag)
    }

  private def startLoginFlow(): Unit =
    activity.showFragment(SignInFragment(SignInMethod(Login, Email)), SignInFragment.Tag)

  private def startEnterpriseLoginFlow(): Unit = extractTokenAndShowSSODialog(showIfNoToken = true)
}

object WelcomeFragment {
  val Tag: String = "WelcomeFragment"

  def apply() = new WelcomeFragment
}
